test/hotspot/jtreg/compiler/loopopts/superword/TestCyclicDependency.java

```diff
@@ -0,0 +1,407 @@
+/*
+ * Copyright (c) 2023, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+/*
+ * @test
+ * @bug 8298935
+ * @summary Writing forward on array creates cyclic dependency
+ *          which leads to wrong result, when ignored.
+ * @requires vm.compiler2.enabled
+ * @library /test/lib /
+ * @run driver TestCyclicDependency
```

Comment:
Yes, it would be interesting to test the flags `SuperWordMaxVectorSize`, `AlignVector` and CompileCommand Directive `Vectorize` in combination (activated / deactivated).

Reply:
Sounds good.

Reply:
@jatin-bhateja I am trying to script-generate lots of test functions, and the `IR` rules. But this is a bit tricky.

I'm using types `byte, char, short, int, long, float, double` with offsets in a range from `-129 .... 0 ... 129`.

I need to ensure that we have enough elements per vector. For that, I need to ensure the loop is unrolled sufficiently. Even if I set `LoopMaxUnroll` extremely high, we do often stop unrolling before what I expected.
In `SuperWord::unrolling_analysis` we figure out how many elements of a type fit into a vector, and that sets the `slp_max_unroll`. For that, we ask `Matcher::vector_width_in_bytes(BasicType bt)`. And this limited by `SuperWordMaxVectorSize`, but also depends more on hardware feature like `UseAVX` or `VM_Version::supports_avx512bw`.
For the arm platforms arm32 and aarch64 it seems a bit easier, it just depends on `MaxVectorSize`.

I fear that I will have to make basically a `IR` rule per `UseAVX` value. Maybe for arm platforms its a bit easier.

Please let me know if you have a better idea @jatin-bhateja !

Reply:
@jatin-bhateja I now have a first version out of the `test/hotspot/jtreg/compiler/loopopts/superword/TestDependencyOffsets.java`. It seems to work for `sse4.1 ... avx512`. I'm now testing it for `asimd`.
And then I will proceed to add the features for `-XX:+AlignVector`, with the modulo check.

Reply:
Update: I cannot use `SuperWordMaxVectorSize`, because it is regulated down on x86 to be at most `MaxVectorSize`. But on `aarch64` the flag `SuperWordMaxVectorSize` does not seem to be adjusted. @fg1417 do you think this is correct / on purpose? Maybe this is just an unfortunate but harmless inconsistency. I guess in `SuperWord::max_vector_size` we first get the info from `Matcher::max_vector_size` (based on `MaxVectorSize`), and then upper bound that based on `SuperWordMaxVectorSize`.

Reply:
Hi @eme64, see https://github.com/openjdk/jdk/pull/8877. Before that, we use `MaxVectorSize` for all platforms.  `SuperWordMaxVectorSize` is only used to fix the performance issue on x86. The option is set as `64` by default, which is fine for current aarch64 hardware, but SVE architecture supports more than 512 bits. I believe `SuperWordMaxVectorSize` is just a temporary solution and we expect a more complex fix. Thanks.

Reply:
> @jatin-bhateja I now have a first version out of the `test/hotspot/jtreg/compiler/loopopts/superword/TestDependencyOffsets.java`. It seems to work for `sse4.1 ... avx512`. I'm now testing it for `asimd`. And then I will proceed to add the features for `-XX:+AlignVector`, with the modulo check.

I am seeing lots of IR violations with UseSSE=4.

Reply:
I realized I have lots of negative IR rules that check that we do NOT vectorize if I expect cyclic dependency. But these negative rules are difficult, there may always be some other factor that leads to shorter vector sizes than what I expect. And then it vectorizes, and does not encounter a cyclic dependency. So I will have to remove all these negative IR rules.

Reply:
I now removed all such negative IR rules.

Reply:
Hi @eme64 ,
Few more suggestions, my original request was to write few hand crafted tests with combinations of AlignVector, Vectorize and MaxVectorSize options. All the test points in TestDependencyOffsets.java are mainly around two kernels, forward writes (RAW/true dependency) and forward reads (WAR/anti-dependence) but exhaustively generates multiple IR rules.
applyIfCPUFeature uses VM identified CPU feature checks which are constrained by UseSSE and UseAVX options. You may extend your script to generate curated scenarios  like following to trigger more rules.

```
        // CPU: sse4.1 to avx -> vector_width: 16 -> elements in vector: 4
        scenarios[0] = new Scenario(0, "-XX:-/+AlignVector", "-XX:MaxVectorSize=XXX", "-XX:UseAVX=0");
        // CPU: avx2 -> vector_width: 32 -> elements in vector: 8
        scenarios[1] = new Scenario(1, "-XX:-/+AlignVector", "-XX:MaxVectorSize=XXX", "-XX:UseAVX=2");
        // CPU: avx512 -> vector_width: 64 -> elements in vector: 16
        scenarios[2] = new Scenario(2, "-XX:-/+AlignVector", "-XX:MaxVectorSize=XXX", "-XX:UseAVX=3");
```
 
Since there are mainly 4 categories of features.

```
CPROMPT>grep "applyIfCPUFeatureAnd" TestDependencyOffsets.java | sort -u | uniq
        applyIfCPUFeatureAnd = {"avx2", "true", "avx512bw", "false"})
        applyIfCPUFeatureAnd = {"avx2", "true", "avx512", "false"})
        applyIfCPUFeatureAnd = {"avx", "true", "avx512", "false"})
        applyIfCPUFeatureAnd = {"sse4.1", "true", "avx2", "false"})
        applyIfCPUFeatureAnd = {"sse4.1", "true", "avx", "false"})
```
 
Also, this test file is script generated hence we may generate separate test files for AARCH64 and X86 to make the tests more maintainable.

Reply:
Hi @jatin-bhateja .
At Oracle we run the `compiler/loopopts/superword` and `compiler/vectorization` tests with various `AVX` and `SSE` settings, including the `UseKNLSetting`.

I could remove the exhaustive loop over all `MaxVectorSize` values, and instead run only a few per `AVX / SSE` setting. That may make it a bit more efficient, and easier to test everything at once on one machine.

I can also make one file `x86` specific (require `x86/x64`) and the other file allows all other platforms (but only contains the `asimd` IR rules). I would probably have different scenarios for that second file, since ARM chips have various `MaxVectorSize`, depending on hardware.

Reply:
Agree, it will be better if we can run multiple IR rules on one target using UseSSE/UseAVX flags rather than having tight target dependencies.

Reply:
@jatin-bhateja do your agree with the recent changes to TestDependencyOffsets.java?
Do you think the jtreg-tests are ok, with their requirements?